src/audio/asrc/asrc.c

```diff
@@ -487,26 +598,110 @@ static int asrc_prepare(struct comp_dev *dev)
 		goto err_free_asrc;
 	}
 
-	ret = asrc_update_drift(cd->asrc_obj, cd->drift);
+	/* Prefer previous skew factor. If the component has not yet been
+	 * run the skew is zero from new(). In that case use factor 1.0
+	 * to start with.
+	 */
+	if (!cd->skew)
+		cd->skew = Q_CONVERT_FLOAT(1.0, 30);
+
+	trace_asrc_with_ids(dev, "asrc_prepare(), skew = %d", cd->skew);
+	ret = asrc_update_drift(cd->asrc_obj, cd->skew);
 	if (ret) {
 		trace_asrc_error_with_ids(dev, "asrc_update_drift(), error %d",
 					  ret);
 		goto err_free_asrc;
 	}
 
+	/* Enable timestamping in pipeline DAI */
+	if (cd->track_drift) {
+		ret = asrc_dai_find(dev, cd, sinkb, sourceb);
+		if (ret) {
+			trace_asrc_error_with_ids(dev, "No DAI found to track");
+			cd->track_drift = false;
+			goto err_free_asrc;
+		}
+
+		cd->ts_count = 0;
+		asrc_dai_configure_timestamp(cd);
+	}
+
 	return 0;
 
 err_free_asrc:
 	rfree(cd->asrc_obj);
+	cd->asrc_obj = NULL;
 
 err_free_buf:
 	rfree(cd->buf);
+	cd->buf = NULL;
 
 err:
 	comp_set_state(dev, COMP_TRIGGER_RESET);
 	return ret;
 }
 
+static int asrc_control_loop(struct comp_data *cd)
+{
+	struct sof_timestamp_data tsd;
+	int64_t tmp;
+	int32_t delta_sample;
+	int32_t delta_ts;
+	int32_t sample;
+	int32_t ts;
+	int32_t skew;
+	int32_t f_ds_dt;
+	int32_t f_ck_fs;
+	int ts_ret = -1;
+
+	if (cd->track_drift) {
+		ts_ret = asrc_dai_get_timestamp(cd, &tsd);
+		asrc_dai_start_timestamp(cd);
+	}
+
+	if (!ts_ret) {
+		ts = (int32_t)(tsd.walclk); /* Let it wrap, diff unwraps */
```

Comment:
You can save one indentation level if you do the if (ts_ret) return 0; first

Reply:
Yep

Reply:
I noticed the possible error from asrc_dai_get_timestamp() is not propagated. I'll fix it too.

Reply:
Done.

This created a bit more change into ASRC like need like start of timestamp at prepare() and increase TS_STABLE_DIFF_COUNT by one for expected sane sample and clock deltas. It prevents the low-pass filter from being fed severely off values and cause xruns problem.